### PR TITLE
CHAM-586 Allow numbers and symbols on yaml context

### PR DESCRIPTION
**Pull request by nkarab**
Problem and/or solution
-----------------------
So far, we've been able to include context in YAML files (YAML_GENERIC, YML i18n) [tx doc](https://help.transifex.com/en/articles/6223189-yaml#h_b26ad4b4a4) by utilizing custom [YAML tags](https://yaml.org/spec/1.2.2/#node-tags) .
 For instance:

```yaml
string: !context translation 
```
However, our Yaml parser for the context only allows alphabetic characters, zeros, and underscores. For instance:

```yaml
a_string_without_context: !context1 translation  # context -> None
a_string_with_context: !context_smth translations  # context -> !context_smth
```
We have now extended support to include numbers and a limited set of special symbols namely (- _ : .).

Additionally, the context is stored in our database with the exclamation mark, and we aim to remove it.

**backwards compatibility**
slack conversation about it: https://transifex.slack.com/archives/C01BHGRAGH1/p1693309705338759

How to test
-----------
Upload a YML i18n and a YAML_GENERIC resource file that includes custom tags featuring numbers or special symbols.

```yaml
en:
  context_string: !he:fd94_fd-la. "context string"
  verbim_context_string: !<!context545qa> "verbim context string"
  context_on_nested_map:
    first: !first_context:54KJFLA95KJ4 "context in nested map"
    second: !second_context:FDKJ40DK "context in nested map"
```

We expect these contextes:

string key | context
-- | --
context_string | he:fd94_fd-la.
verbim_context_string | context545qa
context_on_nested_map.first | first_context:54KJFLA95KJ4
context_on_nested_map.second | second_context:FDKJ40DK

We expect a downloaded file for the source language like this:
```yaml
en:
  context_string: !he:fd94_fd-la. "context string"
  verbim_context_string: !context545qa "verbim context string"
  context_on_nested_map:
    first: !first_context:54KJFLA95KJ4 "context in nested map"
    second: !second_context:FDKJ40DK "context in nested map"
```

We expect for the translation language a file like this: 
```yaml
el:
  context_string: !he:fd94_fd-la. ""
  verbim_context_string: !contex545qa ""
  context_on_nested_map:
    first: !first_context:54KJFLA95KJ4 ""
    second: !second_context:FDKJ40DK ""
    
```

**Staging**: https://yamlcntxt.stg.transifex.net/

Reviewer checklist
------------------

Code:
* [ ] Change is covered by unit-tests
* [ ] Code is well documented, well styled and is following [best practices](https://tem.transifex.com)
* [ ] Performance issues have been taken under consideration
* [ ] Errors and other edge-cases are handled properly

PR:
* [ ] Problem and/or solution are well-explained
* [ ] Commits have been squashed so that each one has a clear purpose
* [ ] Commits have a proper commit message [according to TEM](https://tem.transifex.com/github-guide.html#working-on-a-feature)

### CODE REVIEW

**Review by fathineos — COMMENTED** [review]
Let's add a test case with `!!` as well

**Review by kbairak — DISMISSED** [review]

File: openformats/formats/yaml/utils.py [hunk 1]
```diff
@@ -142,16 +142,22 @@ def _is_custom_tag(self, tag):
...
+
         Built-in types, indicated by a `!!` prefix, will not be matched. We
         can't preserve the information whether a built-in tag like `!!str` was
         used for a value since the PyYAML library will tag such entries with
         the built-in identifier. For example `tag:yaml.org,2002:str`, not
         `!!str`.
         """
-        return re.match(ensure_unicode(r'^[\![a-zA-Z_]*]*$'),
+
+        return re.match(ensure_unicode(r'^\![a-zA-Z0-9_:.\-]*$'),
```
Comment (kbairak) [inline openformats/formats/yaml/utils.py:157]: Why not `\w` instead of `a-zA-Z0-9`?

File: openformats/formats/yaml/utils.py [hunk 2]
```diff
@@ -142,16 +142,22 @@ def _is_custom_tag(self, tag):
...
                         tag,
                         re.IGNORECASE)
-
+
```
Comment (kbairak) [inline openformats/formats/yaml/utils.py:160]: The linter is not going to like this.

**Review by kbairak — APPROVED** [review]
Looks great! The comment is optional.

**Review by kbairak — APPROVED** [review]
